File: python/mlmorph/foreign_word_detector.py
#!/usr/bin/env python

import re

ENGLISH_PATTERNS = [
    "\\S+[അ-ഔ]\\S+",
    "\\S+റ്റ്\\S+",
    "^ജെ",
    "^ട",
    "^ഡ",
    "^ഫാ",
    "^ഫി",
    "^ഫീ",
    "^ഫു",
    "^ഫെ",
    "^ഫെ",
    "^ഫൈ",
    "^ഫൊ",
    "^ഫോ",
    "^ബാൻ",
    "^ബാറ്റ്",
    "^ബെ",
    "^മെറ്റ",
    "മ്യൂ",
    "^ഷോ",
    "^സ്ക്രീ",
    "^ഹാ",
    "^ഹി",
    "^ഹെ",
    "^ഹൊ",
    "^ഹോ",
    "^ഹൗ",
    "^ഹ്യ",
    "^ഹ്വ",
    "^ള",
    "^ഴ",
    "^റ",
    "^റോ",
    "^റ്റൈ",
    "ഓൺ",
    "ക്ച",
    "ക്ട്$",
    "ക്യൂ",
    "ക്സ",
    "ക്റ്റ്",
    "ഗ്രഫ",
    "ഗ്രഫി",
    "ങ്സ്",
    "ച്വ",
    "ജ്യു",
    "ജ്യൂ",
    "ട്ര",
    "ഡിയോ",
    "ഡ്$",
    "ഡ്ക",
    "ഡ്മ",
    "ഡ്വ",
    "ഡ്സ",
    "ണിയൻ$",
    "ണിയർ$",
    "ണ്ടർ$",
    "ൻഡ",
    "ൻസ്",
    "ന്റം",
    "ന്റിക്",
    "ന്റ്$",
    "ന്റ്സ്",
    "ൻ്റ്$",
    "ന്റർ",
    "പ്രൊ",
    "പ്സ്$",
    "പ്റ്റ",
    "ഫ്ര",
    "ഫ്ല",
    "ഫ്ള",
    "ഫ്റ്റ",
    "ഫൈ",
    "ബിൻ",
    "ബിർ",
    "ബിൽ",
    "ബിൾ",
    "ബിറ്റ്",
    "ബെർ",
    "ബൈ",
    "ബ്രേ",
    "ബ്രോ",
    "ബ്ല",
    "ബ്സ",
    "ബ്ള",
    "മെമ്മ",
    "മൈസ",
    "മ്പ്യൂ",
    "യിറ്റ്",
    "യ്ൻ",
    "യ്സ",
    "യ്സ്",
    "യിസ",
    "യ്റ്റ്",
    "ർജ",
    "ർട",
    "ലിറ്റി",
    "ലീസ്",
    "ലൈസർ",
    "ൽസ്",
    "വെയർ$",
    "വൈസ്",
    "വേസ്",
    "ഷിസ്",
    "ഷ്യൻ$",
    "ഷ്യർ$",
    "ഷ്യൽ$",
    "ഷ്യസ്$",
    "സബ്",
    "സർ$",
    "സൽ$",
    "സെൽ",
    "സെർ",
    "സെൻ",
    "സിസ്",
    "സിറ്റി",
    "സെന്റ",
    "സ്കൂൾ",
    "സ്പേസ്",
    "സൈറ്റ",
    "സ്പ്ല",
    "സ്പ്ള",
    "സ്ബ",
    "സ്ല",
    "സ്ള",
    "സ്കാൻ",
    "സ്കോ",
    "സ്റ്റ",
    "^സ്വീ",
    "സ്ല",
    "സ്ട്ര",
    "ളജി",
    "ൾട്ട",
    "ഴ്സ",
    "റ്റിക്$",
    "റ്റിക്കലി$",
    "റ്റേൺ",
    "റ്ററ",
    "റ്റർ",
    "റ്റലൈ",
    "റൈറ്റ",
    "ിംഗ്$",
    "ിങ്$",
    "ിങ്ങ്$",
    "ിഷൻ",
    "ിസം$",
    "ിംസ്",
    "ീസ്$",
    "േഷൻ",
    "ൈറ്റ",
    "ോഷൻ",
    "ൗണ്ട",
    "്രീം",
    "ക്വേ"
]

compiled_english_pattern_regex = re.compile("|".join(ENGLISH_PATTERNS))


def check_foreign_word(word: str) -> int:
    """
    Check if the word is foreign word or not

    Args
    ----
        word : str, word to check whether it is foreign word or not

    Returns
    -------
        int : A score between 0 and 1
    """
    word = word.strip()
    # Remove all ZWS, ZWNJ, ZWJ before pattern matching
    word = re.sub(r"[\u200B-\u200D]", "", word)
    if not is_valid_malayalam_word(word):
        # Unknown word. Surely foreign
        return 1
    if has_sure_patterns(word):
        return 1
    return 0


def has_sure_patterns(word: str) -> bool:
    return re.search(compiled_english_pattern_regex, word)


def is_valid_malayalam_word(word: str) -> bool:
    if len(word) <= 1:
        return False
    # Ignore all non-Malayalam words
    if re.search(r"[\u0D00-\u0D7F\u200C-\u200D]+", word) is None:
        return False
    return True

File: rust/src/foreign_word_detector.rs
use regex::Regex;

const ENGLISH_PATTERNS: &[&str] = &[
    r"\\S+[അ-ഔ]\\S+",
    r"\\S+റ്റ്\\S+",
    r"^ജെ",
    r"^ട",
    r"^ഡ",
    r"^ഫാ",
    r"^ഫി",
    r"^ഫീ",
    r"^ഫു",
    r"^ഫെ",
    r"^ഫെ",
    r"^ഫൈ",
    r"^ഫൊ",
    r"^ഫോ",
    r"^ബാൻ",
    r"^ബാറ്റ്",
    r"^ബെ",
    r"^മെറ്റ",
    r"മ്യൂ",
    r"^ഷോ",
    r"^സ്ക്രീ",
    r"^ഹാ",
    r"^ഹി",
    r"^ഹെ",
    r"^ഹൊ",
    r"^ഹോ",
    r"^ഹൗ",
    r"^ഹ്യ",
    r"^ഹ്വ",
    r"^ള",
    r"^ഴ",
    r"^റ",
    r"^റോ",
    r"^റ്റൈ",
    r"ഓൺ",
    r"ക്ച",
    r"ക്ട്$",
    r"ക്യൂ",
    r"ക്സ",
    r"ക്ലോ",
    r"ക്റ്റ്",
    r"ഗ്രഫ",
    r"ഗ്രഫി",
    r"ഗിൾ",
    r"ങ്സ്",
    r"ച്വ",
    r"ജ്യു",
    r"ജ്യൂ",
    r"ട്ര",
    r"ഡിയോ",
    r"ഡ്$",
    r"ഡ്ക",
    r"ഡ്മ",
    r"ഡ്വ",
    r"ഡ്സ",
    r"ണിയൻ$",
    r"ണിയർ$",
    r"ണ്ടർ$",
    r"ൻഡ",
    r"ൻസ്",
    r"ന്റം",
    r"ന്റിക്",
    r"ന്റ്$",
    r"ന്റ്സ്",
    r"ൻ്റ്$",
    r"ന്റർ",
    r"പ്രൊ",
    r"പ്സ്$",
    r"പ്റ്റ",
    r"ഫ്ര",
    r"ഫ്ല",
    r"ഫ്ള",
    r"ഫ്റ്റ",
    r"ഫൈ",
    r"ബിൻ",
    r"ബിർ",
    r"ബിൽ",
    r"ബിൾ",
    r"ബിറ്റ്",
    r"ബെർ",
    r"ബൈ",
    r"ബ്രേ",
    r"ബ്രോ",
    r"ബ്ല",
    r"ബ്സ",
    r"ബ്ള",
    r"മെമ്മ",
    r"മൈസ",
    r"മ്പ്യൂ",
    r"യിറ്റ്",
    r"യ്ൻ",
    r"യ്സ",
    r"യ്സ്",
    r"യിസ",
    r"യ്റ്റ്",
    r"ർജ",
    r"ർട",
    r"ലിറ്റി",
    r"ലീസ്",
    r"ലൈസർ",
    r"ൽസ്",
    r"വെയർ$",
    r"വൈസ്",
    r"വേസ്",
    r"ഷിസ്",
    r"ഷ്യൻ$",
    r"ഷ്യർ$",
    r"ഷ്യൽ$",
    r"ഷ്യസ്$",
    r"സബ്",
    r"സർ$",
    r"സൽ$",
    r"സെൽ",
    r"സെർ",
    r"സെൻ",
    r"സിസ്",
    r"സിറ്റി",
    r"സെന്റ",
    r"സ്കൂൾ",
    r"സ്പേസ്",
    r"സൈറ്റ",
    r"സ്പ്ല",
    r"സ്പ്ള",
    r"സ്ബ",
    r"സ്ല",
    r"സ്ള",
    r"സ്കാൻ",
    r"സ്കോ",
    r"സ്റ്റ",
    r"^സ്വീ",
    r"സ്ല",
    r"സ്ട്ര",
    r"ളജി",
    r"ൾട്ട",
    r"ഴ്സ",
    r"റ്റിക്$",
    r"റ്റിക്കലി$",
    r"റ്റേൺ",
    r"റ്ററ",
    r"റ്റർ",
    r"റ്റലൈ",
    r"റൈറ്റ",
    r"ിംഗ്$",
    r"ിങ്$",
    r"ിങ്ങ്$",
    r"ിഷൻ",
    r"ിസം$",
    r"ിംസ്",
    r"ീസ്$",
    r"േഷൻ",
    r"ൈറ്റ",
    r"ോഷൻ",
    r"ൗണ്ട",
    r"്രീം",
    r"ക്വേ",
    r"[0-9]+",
];

lazy_static::lazy_static! {
    static ref COMPILED_ENGLISH_PATTERN: Regex = {
        let pattern = ENGLISH_PATTERNS.join("|");
        Regex::new(&pattern).unwrap()
    };

    static ref MALAYALAM_PATTERN: Regex = {
        Regex::new(r"[\u{0D00}-\u{0D7F}\u{200C}-\u{200D}]+").unwrap()
    };

    static ref ZWS_PATTERN: Regex = {
        Regex::new(r"[\u{200B}-\u{200D}]").unwrap()
    };
}

pub fn check_foreign_word(word: &str) -> i32 {
    let word = word.trim();
    // Remove all ZWS, ZWNJ, ZWJ before pattern matching
    let word = ZWS_PATTERN.replace_all(word, "");

    if !is_valid_malayalam_word(&word) {
        // Unknown word. Surely foreign
        return 1;
    }

    if has_sure_patterns(&word) {
        return 1;
    }

    0
}

fn has_sure_patterns(word: &str) -> bool {
    COMPILED_ENGLISH_PATTERN.is_match(word)
}

fn is_valid_malayalam_word(word: &str) -> bool {
    if word.len() <= 1 {
        return false;
    }

    // Ignore all non-Malayalam words
    MALAYALAM_PATTERN.is_match(word)
}
